### PR TITLE
prepare mypy==1.13.0: legacy cast

### PR DESCRIPTION
**Pull request by JakobSpahn**
Use a helper for better readability compared to #7846. Two different calls to a `cast` in one line is confusing.

### CODE REVIEW

**Comment by JakobSpahn** [comment]
Other than throwing `# type: ignore` I don't see how we can avoid narrowing the str to the literal types of the [format parameters in the stub](https://github.com/python/mypy/blob/v1.13.0/mypy/typeshed/stdlib/builtins.pyi#L769).

File: tinygrad/helpers.py [hunk 1]
```diff
@@ -280,6 +280,12 @@ class CStruct(ctypes.Structure):
 def init_c_var(ctypes_var, creat_cb): return (creat_cb(ctypes_var), ctypes_var)[1]
 def flat_mv(mv:memoryview): return mv if len(mv) == 0 else mv.cast("B", shape=(mv.nbytes,))
 
+# TODO typeshed stub is missing the stub for memoryview float16 cast (format: Literal['e', '@e']). Remove once support is added.
+# src: https://github.com/python/mypy/blob/v1.13.0/mypy/typeshed/stdlib/builtins.pyi#L769
+def to_legacy_fmt(fmt:str):
```
Comment (geohot) [inline tinygrad/helpers.py:285]: I don't think this is the right way to handle this. Think about what you want to do with this, not a noop helper

**Comment by JakobSpahn** [comment]
Switched to narrowing the type via asserts instead of `typing.cast`. Is a few more lines than the other, but more readable imo.

File: tinygrad/runtime/ops_python.py [hunk 2]
```diff
@@ -68,9 +69,11 @@ def __call__(self, *bufs, global_size:Tuple[int,int,int]=(1,1,1), local_size:Tup
         dl[i] = dtype
         if uop is Ops.DEFINE_GLOBAL:
           assert dtype.fmt is not None
+          if TYPE_CHECKING or sys.version_info < (3, 12): assert dtype.fmt != "e"
```
Comment (geohot) [inline tinygrad/runtime/ops_python.py:72]: Don't copy paste, refactor to be clean

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -286,6 +287,7 @@ def item(self) -> ConstType:
     """
     assert self.dtype.fmt is not None, f"no fmt dtype for {self.dtype}"
     assert self.numel() == 1, "must have one element for item"
+    if TYPE_CHECKING or sys.version_info < (3, 12): assert self.dtype.fmt != "e"
```
Comment (geohot) [inline tinygrad/tensor.py:290]: Don't copy paste. This function should probably just call `data()`

**Comment by geohot** [comment]
Close

**Comment by geohot** [comment]
Why is this draft? Looks good

**Pull request merged**